### PR TITLE
test: upgrade storybook to v7, use vite

### PR DESCRIPTION
**Pull request by lesterchoi-okta**
### Description

package `@okta/odyssey-storybook`:

- upgrades `storybook` and related dependencies to `^7.0.7` (from `^6.15.5`)
- change storybook builder from `webpack4` to `@storybook/react-vite`
- fixes "Dynamic title" error caused by slant quotes (breaking change in `storybook@7`)
- replace deprecated type `Story` to `StoryFn` and `StoryObj`
- replace deprecated type `ComponentMeta` to `Meta`

package `@okta/odyssey-react-mui`:

- export "./README.md" to import in storybook

package `@okta/odyssey-react-labs`:

- export "./README.md" to import in storybook

### CODE REVIEW

File: packages/odyssey-storybook/src/installation/Odyssey-React-MUI.stories.mdx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { Meta } from "@storybook/addon-docs/blocks";
-
-import Readme from "@okta/odyssey-react-mui/README.md";
+import Markdown from "markdown-to-jsx";
+import Readme from "@okta/odyssey-react-mui/README.md?raw";
```
Comment (lesterchoi-okta) [inline packages/odyssey-storybook/src/installation/Odyssey-React-MUI.stories.mdx:3]: requires "./README.md" to be to exported from `packages/odysssey-react-mui/package.json`

File: packages/odyssey-react-labs/package.json [hunk 2]
```diff
@@ -12,6 +12,9 @@
     ".": {
       "default": "./dist/index.js",
       "types": "./dist/src/index.d.ts"
+    },
+    "./README.md": {
+      "default": "./README.md"
     }
```
Comment (lesterchoi-okta) [inline packages/odyssey-react-labs/package.json:18]: required to import from `packages/odyssey-storybook/src/installation/Odyssey-React-MUI.stories.mdx`

File: packages/odyssey-react-mui/package.json [hunk 3]
```diff
@@ -12,6 +12,9 @@
     ".": {
       "types": "./dist/src/index.d.ts",
       "default": "./dist/index.js"
+    },
+    "./README.md": {
+      "default": "./README.md"
     }
```
Comment (lesterchoi-okta) [inline packages/odyssey-react-mui/package.json:18]: required to import from `packages/odyssey-storybook/src/installation/Odyssey-React-MUI.stories.mdx`

File: packages/odyssey-storybook/.storybook/preview.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        ],
+        locales: "",
+      },
+    },
+    viewMode: "docs",
+    previewTabs: {
+      "storybook/docs/panel": { index: -1 },
+    },
+    backgrounds: {
+      default: "white",
```
Comment (lesterchoi-okta) [inline packages/odyssey-storybook/.storybook/preview.ts:35]: note: storybook logs warning re; invalid name "white"

File: packages/odyssey-storybook/vite.config.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+/* eslint-disable import/no-extraneous-dependencies */
+import { defineConfig } from "vite";
+import react from "@vitejs/plugin-react-swc";
+
+// https://vitejs.dev/config/
+export default defineConfig({
+  plugins: [react()],
+  // assetsInclude: [
+  //   '**/*.md'
+  // ]
```
Comment (lesterchoi-okta) [inline packages/odyssey-storybook/vite.config.ts:22]: nit: remove dead code

**Review by lesterchoi-okta — COMMENTED** [review]
1.  seems like storybook removed support for gfm (github flavored markdown)
2. `mdx` are not being detected as stories, i.e., showing up in a separate subtree from their corresponding stories

**Comment by lesterchoi-okta** [comment]
There's something wrong with the mdx files. Looking into it now.

```
TypeError: import_ts_dedent4.default is not a function
```

**Review by KevinGhadyani-Okta — COMMENTED** [review]
Few early comments. Still going through it.

**Review by KevinGhadyani-Okta — COMMENTED** [review]
Few early comments. Still going through it.

**Review by KevinGhadyani-Okta — APPROVED** [review]

**Review by KevinGhadyani-Okta — APPROVED** [review]

**Comment by edburyenegren-okta** [comment]
Seeing a few issues when testing locally:

1. Docs pages aren't rendering, and I don't see any way to get to them. Am I missing something obvious? 😅 
2. `Button` stories are broken; maybe others
3. `defaultValue` isn't being passed in anywhere. It wasn't deprecated in SB7, so I'm not sure what's wrong: ﻿https://storybook.js.org/docs/react/api/argtypes

**Comment by lesterchoi-okta** [comment]
> Docs pages aren't rendering, and I don't see any way to get to them. Am I missing something obvious? 😅

Do you mean something was (re)moved from the Storybook UI? I see docs pages in the navigation

localhost_6006__path=_docs_mui-components-dialog--docs(Desktop (1440×900)) (1)


> Button stories are broken; maybe others

Could you clarify what you mean by "broken"? They seem to be working for me.

> defaultValue isn't being passed in anywhere. It wasn't deprecated in SB7, so I'm not sure what's wrong: ﻿https://storybook.js.org/docs/react/api/argtypes
Edit: followup.
> Ignore (3) above. [defaultValue has been deprecated](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#no-longer-inferring-default-values-of-args); they just haven't updated their docs.
> https://github.com/storybookjs/storybook/issues/11983#issuecomment-896790639, but some people have reported that this doesn't actually change the render. Instead, it sets the value inside the controls table.
> I found a possible solution where [we set the args up front](https://github.com/yobgob/too-many-hooks/pull/50/files) that we want the stories to share.

SB FAQ: "[Why are my args no longer displaying the default values?](https://storybook.js.org/docs/react/faq#why-are-my-args-no-longer-displaying-the-default-values)"

I'd suggest using the spread operator as you see fit to set args at the Story level

```tsx
const defaultArgs: ButtonProps = {
  text: 'Click me',
  variant: 'primary',
}

// option 1: set on a template (not recommended)
const ButtonTemplate: StoryObj<ButtonProps> = {
  args: defaultArgs, 
  play: async () => {}
}
export const ButtonPrimary: StoryObj<ButtonProps> = {
  ...ButtonTemplate, // spreads "play" and "args" into story
  // args: { text: 'Oops' } // variant would be undefined if this line is enabled
}

// option 2: set on each story
export const ButtonSecondary: StoryObj<ButtonProps> = {
  args: {
    ...defaultArgs, // NOTE: important that defaultArgs precedes override(s) below
    variant: 'secondary'
  }
}

// option 3: use custom merge function (not recommended)
// custom merge is less readable, may encourage unexpected transformations, open door to errors
```

**Comment by edburyenegren-okta** [comment]
Just pulled your latest; most things are working for me now!

> Do you mean something was (re)moved from the Storybook UI? I see docs pages in the navigation

Seeing this now. Has v7 removed the ability to have docs visible on the stories, rather than having a separate "docs" page? Not a huge deal but may be confusing for users who are used to the other behavior.

> Could you clarify what you mean by "broken"? They seem to be working for me.

I was getting an error when attempting to load them. I saw a flash of it on the initial load this time, but not on subsequent refreshes. I'd paste it here, but can't reliably repro. Hopefully just some wonk on my local.

Not a big deal, but do you know why some stories start in a focused state and not others?

Screenshot 2023-05-22 at 9 44 55 AM<br/>

> I'd suggest using the spread operator as you see fit to set args at the Story level

Any reason not to set `args` up front? Seems simpler and less prone to user error if this is okay.

```
const storybookMeta: Meta<TextFieldProps> = {
  title: "MUI Components/Forms/TextField",
  component: TextField,
  argTypes: {
    autoCompleteType: {
      control: "text",
      defaultValue: "name",
    },
    hasInitialFocus: {
      control: "boolean",
      defaultValue: false,
    },
    isDisabled: {
      control: "boolean",
      defaultValue: false,
    },
    endAdornment: {
      control: "text",
    },
    errorMessage: {
      control: "text",
    },
    hint: {
      control: "text",
    },
    id: {
      control: "text",
    },
    label: {
      control: "text",
      defaultValue: "Destination",
    },
   [...]
  },
  args: {
    label: "Destination",
  },
  decorators: [MuiThemeDecorator],
};
```

This seems to be working everywhere I've tried it. I saw it as a recommended solution in some of SB's GH issues about this.

Whatever route we take, we'll need all the stories to look as expected before we can release. Toasts and Form elements seem the most effected right now.

**Comment by lesterchoi-okta** [comment]
> I was getting an error when attempting to load them. I saw a flash of it on the initial load this time, but not on subsequent refreshes. I'd paste it here, but can't reliably repro. Hopefully just some wonk on my local.

I've had that happen too, but it's a dev-time issue -- it should not show up in the built storybook.

> Not a big deal, but do you know why some stories start in a focused state and not others?

Ah, the primary button is like that because it's the end result of the [example interaction test](https://github.com/okta/odyssey/blob/08a01a6803e565a5b2d9f9c312369fc283e4e0e4/packages/odyssey-storybook/src/components/odyssey-mui/Button/Button.stories.tsx#L64-L74)

You can tell by looking at the "actions" and "interactions" tabs in the section below the canvas (not sure what that container's called)

> Any reason not to set args up front? Seems simpler and less prone to user error if this is okay.

I don't think I saw that part of your previous comment. If that works, it does seem cleaner. Didn't know `StorybookMeta` had that property and behavior.

**Pull request merged**